clients/include/utility.h

```diff
@@ -123,54 +145,262 @@ inline float half_to_float(rocblas_half val)
     return _cvtsh_ss(val);
 }
 
+// Random number generator
+typedef mt19937 rocblas_rng_t;
+extern rocblas_rng_t rocblas_rng, rocblas_seed;
+
+// Reset the seed (mainly to ensure repeatability of failures in a given suite)
+inline void rocblas_seedrand()
+{
+    rocblas_rng = rocblas_seed;
+}
+
+/* ============================================================================================ */
+/*! \brief  Memory guard detects memory coruption */
+template <class T, size_t PAD>
+class memory_guard
+{
+    T guard[PAD];
+
+    // Generate random NaN values
+    template <typename U, typename UINT_T, int SIG, int EXP>
+    static U random_nan_data()
+    {
+        static_assert(sizeof(UINT_T) == sizeof(U), "Type sizes do not match");
+        union
+        {
+            UINT_T u;
+            U fp;
+        } x;
+        do
+            x.u = uniform_int_distribution<UINT_T>()(rocblas_rng);
+        while(!(x.u & (((UINT_T)1 << SIG) - 1))); // Reject Inf (mantissa == 0)
```

Comment:
If this is slowing down the code could you create a     pad_mask  in random_data . There would be a separate pad_mask for half, float, double, and you could  replace the two lines with
x.u  |=  pad_mask;
The pad_mask would set the exponent to all ones and set the lowest mantissa bit to 1

Reply:
It shouldn't be slowing down the code; it just does one `test` instruction (the compiler optimizes the constant mask expression) and a branch if the `(value & mask) == 0`. There's a `1/2^SIG` probability that it repeats. Repeating it ensures that all legal values are equally likely. Setting the lowest mantissa `== 1` all the time, would throw away half of the possible random values. There is no need for a separate `pad_mask` because the compiler optimizes constant expressions, and `SIG` and `EXP` are constants. This is the assembly code for the test and mask above:

```
        movabs  rbp, 4503599627370495        ; 2^SIG-1

.L31:
...     
     ;  rax = uniform_int_distribution<UINT_T>()(rocblas_rng)

        test    rax, rbp   ; Tests whether mantissa == 0
        je      .L31       ; Only repeats if mantissa == 0; 1/2^SIG probability
        movabs  rbx, 9218868437227405312   ; (2^EXP-1)<<SIG
        or      rax, rbx   ; Sets all exponent bits to 1
...
 ```

It tests whether `%rax` ANDed with `2^SIG-1` is zero, and if it is, gets another random number (only happens with `1/2^SIG` chance). It then ORs to set the exponent to all 1's.

The expected number of iterations is `1/(1 - 2^-SIG)`, which is:
```
double: 4503599627370496 / 4503599627370495 = 1.0000000000000002...
float:  8388608 / 8388607                   = 1.000000112...
half:   1024 / 1023                         = 1.001...
```

In practice, the `TEST` and `JE` instructions take 0 cycles, because they overlap other instructions which execute out-of-order.